### ISSUE TITLE
Understanding the code

### ISSUE THREAD

**Issue by appleleaves** [larq/larq#176]
1. In larq/layers_base.py line 40.
I could not find the definition  of self.kernel.

2. Also, as I understand, the code quantilize a tensor using utils.quantize() and set the quantilized tensor as the weights. The remained processes such as forward and backward apply the keras original function to do the work, right?

**Comment by lgeiger** [comment]
Thanks for taking the time to look into the code of `larq`.

The general execution flow of the quantized layers is described at https://larq.dev/guides/key-concepts/#quantized-layers

The implementation contains a bit of magic so we don't need to reimplement all `tf.keras` layers and so we can support multiple version while automatically including all performance optimizations from TensorFlow.

A quantized layer inherits from the `larq` base class and the upstream `tf.keras` layer that implements the [forward pass](https://github.com/tensorflow/tensorflow/blob/d394f61b529c44e9b7a3aed230993c0138c8b4a4/tensorflow/python/keras/layers/core.py#L1035-L1055) and [defines all variables](https://github.com/tensorflow/tensorflow/blob/d394f61b529c44e9b7a3aed230993c0138c8b4a4/tensorflow/python/keras/layers/core.py#L1002-L1033):
https://github.com/larq/larq/blob/ed698187e8e4254afef5750403b3be946d98b8f3/larq/layers.py#L19

The `larq` base class overwrites the forward pass by first quantizing the input:
https://github.com/larq/larq/blob/ed698187e8e4254afef5750403b3be946d98b8f3/larq/layers_base.py#L55-L62

Here `self.kernel` is temporarily replaced by the quantized version using the `utils.quantize()` context manager:
https://github.com/larq/larq/blob/ed698187e8e4254afef5750403b3be946d98b8f3/larq/utils.py#L40-L56

This is important since in eager mode other methods like `layers.trainable_variables` rely on the full precision kernel to correctly compute the gradients.

The backward pass is automatically handled by automatic differentiation of TensorFlow and each quantizer defines a custom pseudo gradient to make operations like `tf.sign` differentiable.

Does this answer your questions?

**Comment by lgeiger** [comment]
Closing, due to inactivity. @appleleaves Feel free to reopen if you need more clarification.
